140-runner-host-mounts/proposal.md

```diff
@@ -0,0 +1,75 @@
+
+For instance, to use an AMD GPU within a container, the container needs access to these paths from the host:
+
+  - `/dev/kfd` - GPU compute interface
+  - `/dev/dri` - folder containing an interface for each GPU.
+
+Currently, it is not possible to run these workloads using Concourse.
+
+# Proposal
+
+## Worker configuration option `--allowed-host-mounts`
+
+Add a new configuration option to the worker command, `--allowed-host-mounts`:
+The value should be a regular expression. Any host path that the expression fully matches will be *allowed* to be
+mounted into a task's container if it is requested.
+
+This option can be specified only once, but can use the regex OR operator `|` to match separate paths as required.
+
+## Task configuration option `host_mounts`
+
+Add a new configuration key (`host_mounts`) to the schema for `task` steps, which can be used to request paths from the host to be
+mounted into that task's container.
+
+Formally:
+
+  - `host_mounts` is a list of `host-mount-config`s, defaulting to `[]`
+  - A `host-mount-config` is either:
+    - An object, with the following keys:
+      - `host` (required, non-empty): The desired path to mount from the host
+      - `container` (optional): The mount's location in the container. Defaults to the same as `host`
+      - More options may be added down the line, as required.
+    - A string, of format:
+      - `host:container`, where `host` and `container` are strings not containing `:`, and have meanings as defined above.
+      - or, `host`, leaving `container` to default to `host` as above
+      - This is just a shorthand syntax, similar to the one Docker & Docker Compose uses.
+
+## New behaviour when spawning task containers
+
+If the task's `host_mounts` list is empty, then behaviour is unchanged.
+
+If not, then each element in the list must be validated:
+
+  - The `host` key must be fully matched by the `--allowed-host-mounts` regex.
+    - If the `--allowed-host-mounts` option was not specified, validation fails automatically.
+  - The path referred to by the `host` key must exist from the worker's perspective.
+    - Permissions, writability, etc, are not checked.
+
+If any of the elements do not validate, the task fails with an error message explaining why.
+
+If validation has succeeded, each requested path is mounted into the task's container, using the container runtime's native feature for doing so.
+
+# Open Questions
+
+ - If the worker is itself running in a docker container, the mount will need to also be specified there. This will need to be made clear to users of the feature
+ - Mounting directories from the host opens users up to weird permissions errors, which are often counterintuitive. For instance, group membership is not inherited from the worker user, so if the mounted directory is not world-accessible and its user/group ID doesn't match with what's specified in the container, users will get permission errors.
```

Comment:
Let's move this to the Answered section as well, with the answer being that we document that the onus is on operators to ensure their host mounts have the correct permissions and suggesting some useful defaults.